Data/Generate.cpp

```diff
@@ -130,7 +184,6 @@ generate(unsigned int NumberOfEvents, const ComPWA::Kinematics &Kinematics,
          ComPWA::UniformRealNumberGenerator &RandomGenerator,
```

Comment:
Could we combine the two `generate` functions since they share most of the code? To my opinion we could also remove the first one.
`std::vector<Event> generate(unsigned int NumberOfEvents, const ComPWA::Kinematics &Kinematics,
         const ComPWA::PhaseSpaceEventGenerator &Generator,
         ComPWA::Intensity &Intensity,
         ComPWA::UniformRealNumberGenerator &RandomGenerator)`
and 
`std::vector<ComPWA::Event>
generate(unsigned int NumberOfEvents, const ComPWA::Kinematics &Kinematics,
         ComPWA::UniformRealNumberGenerator &RandomGenerator,
         ComPWA::Intensity &Intensity, const std::vector<ComPWA::Event> &phsp,
         const std::vector<ComPWA::Event> &phspTrue)`

Reply:
You mean to make the phsp samples optional?
I have to think about this a bit. Because of the fixed size of the phsp sample in one case, the control flow of the two algorithms is slightly different.

Reply:
Or probably call the first function by the second one in order to avoid duplicate code.

Reply:
I will give that some thought

Reply:
I cant get around writing a few if else statements when I want to combine these two. As a result the number of saved code lines is not that much.
Somehow this has a bit of code smell and implies that we are going down the wrong hole here. Maybe the concept suggested by @zhang-jingqing of introducing an Event unique ID would be a better solution. I don't understand why this reco and mc sample matching has to be done on the level of the event generation.

Reply:
To properly generate events you need a MC sample from which you know the true phsp position as well as the reconstructed position. When you generate a hit&miss sample you need the true position to evaluate the intensity. But the position that you actually select is the reconstructed position. With that procedure you include detector effects which are simulated in the MC without spoiling the (effective) intensity model.

Reply:
As far as I understand now, you need this for input output studies, right? So you want to create a realistic event sample, that is based on your amplitude model. So you would need to do these steps:
1. generate hit&miss MC event sample
2. reconstruct this sample
3. what else?

Reply:
1. You generate and reconstruct a MC phsp sample using the experiment's software framework.
2. You store for each generated and reconstructed position for each event
3. You use those two sample to generate a realistic MC sample of you amplitude